address: optimistic asset sync

jharveyb opened the pull request:
Fixes #620.

This PR adds support for creating addresses for assets that  a node has not bootstrapped or synced. If the asset info is not found, the node tries to sync asset info from all members of their universe federation. If the asset info is found, this asset is also added to the universe sync config to keep track of future possible issuance.

This helps address the UX changes from switching global sync to opt-in, since you can also make an address for an asset that's a member of an asset group, without having bootstrapped the asset group creation itself.

A related change is that the `QueryAssetRoots` RPC will now try and look up an asset group for a given universe ID, iff the ID specified an asset ID and no other roots were found. This is what allows a user to bootstrap an asset group without knowing the group key.

TODOs:

- [x] extend itest to cover a grouped asset and verify asset universe config changes + sync behavior with the config changes

Roasbeef commented on rpcserver.go:
```diff
@@ -2993,6 +2995,109 @@ func (r *rpcServer) QueryAssetRoots(ctx context.Context,
+		rpcsLog.Debugf("No roots found for asset %v, checking if this "+
+			"asset is part of a group", groupedAssetID.String())
+
+		assetGroup, err := r.cfg.TapAddrBook.QueryAssetGroup(
```
Doing it at this level means another round trip (actually 3x) and set of DB queries each time this path is exercised. Alternatively, we can do this from within the main DB query, as we have the asset group view available. Doing it at that lower level woudl also mean we can test this more effectively, which is something we generally want to improve on now that we have the initial alpha release out.

Roasbeef replied:
Discussed offline that we'll defer the optimization to a future PR, but clean up the call site a bit here with some helper functions.

Roasbeef commented on itest/universe_test.go:
```diff
@@ -497,6 +498,52 @@ func testUniverseFederation(t *harnessTest) {
+
+	waitErr := wait.NoError(func() error {
+		var err error
+		groupUniRoots, err = t.tapd.QueryAssetRoots(
```
Shouldn't this succeed on the very first try? Why do we need `wait.NoError` wrapped around the call.

Roasbeef commented on address/book.go:
```diff
@@ -70,6 +70,20 @@ type QueryParams struct {
+type AssetSyncer interface {
+	// FetchAssetInfo queries the universes in our federation for genesis
+	// and asset group information about the given asset ID.
+	FetchAssetInfo(ctx context.Context, assetID *asset.ID) error
```
IIUC, we just want to attempt to sync it we get an error that we don't know of the asset ID on disk. So we don't need to check to see if we know of it, if we just try and then detect the error when we don't.

jharveyb replied:
Resolved later on? That's what we do with `queryAssetInfo`, intent was to keep `QueryAssetGroup` calls separate from those backend by the universe interfaces, and leave room to turn off this JIT bootstrapping. Also I think sync would fail if we minted the asset and never pushed issuance proofs to any universe in our federation.

Roasbeef commented on address/book.go:
```diff
@@ -70,6 +70,20 @@ type QueryParams struct {
+
+	// UpdateSyncConfig updates the sync config for the given asset so that
+	// we sync future issuance proofs.
+	UpdateSyncConfig(ctx context.Context, assetID *asset.ID,
```
Can use `AssetGroup` here. 

I think we also only care about updating the sync config if the asset itself has a group.

jharveyb replied:
Yeah, I guess an ungrouped asset only ever needs to be synced once and therefore never needs a universe config option? For issuance proofs at least.

Roasbeef commented on tapcfg/server.go:
```diff
@@ -301,6 +294,16 @@ func genServerConfig(cfg *Config, cfgLogger btclog.Logger,
+	addrBook := address.NewBook(address.BookConfig{
+		Store:        tapdbAddrBook,
+		Syncer:       universeFederation,
+		SyncerActive: true,
```
Why's this needed? Can't we just say if the passed interface is nil or not?

Roasbeef replied:
Ah sure agree re being able to opt out of it, in which case, I think the server would just not set `Syncer` at all. Then if that's nil, all the new behavior is skipped.

Roasbeef commented on universe/auto_syncer.go:
```diff
@@ -228,27 +235,16 @@ func (f *FederationEnvoy) syncServerState(ctx context.Context,
-	if err != nil {
-		err := fmt.Errorf("unable to fetch set of universe "+
-			"servers: %v", err)
-		f.reportErr(err)
```
I have a commit that removes this report, otherwise it causes the daemon to shutdown.

Roasbeef commented on universe/auto_syncer.go:
```diff
@@ -531,6 +519,147 @@ func (f *FederationEnvoy) SetAllowPublicAccess() error {
+		return nil
+	}
+
+	// Sync with the federation Universe servers in parallel.
```
Why do we need to sync from all the severs in parallel, when we only want a single asset ID here? 

Instead I think we can just select one (at random?), then try to sync with just that instance.

jharveyb replied:
Intent was that you get the 'latest' asset state, e.x. for an asset group one server may only have record of 5 issuances while the official issuer server has a record of 10.

Also just to mirror what we do for the background sync. I think random single sync may not work well if we expect users to have universes with non-overlapping assets in their federation (one host wallet provider, one explorer, one stablecoin issuer, etc.).

Roasbeef commented on universe/auto_syncer.go:
```diff
@@ -531,6 +519,147 @@ func (f *FederationEnvoy) SetAllowPublicAccess() error {
+
+// UpdateSyncConfig updates the sync config for the given asset to that we sync
+// future issuance proofs.
+func (f *FederationEnvoy) UpdateSyncConfig(ctx context.Context,
```
`EnableAssetSync`? or `EnableSyncForAsset`?

Roasbeef commented on address/book.go:
```diff
@@ -193,15 +194,70 @@ func NewBook(cfg BookConfig) *Book {
+	// Asset lookup failed gracefully; continue to asset lookup using the
+	// AssetSyncer if enabled.
+	case errors.Is(err, ErrAssetGroupUnknown):
+		if !b.cfg.SyncerActive {
```
Why would this ever be false in practice?

jharveyb replied:
Intended as user config option; setting syncer to `nil` on startup would also work.

Roasbeef replied:
Gotcha, I think `nil` here matches how we manage optional interface impls elsewhere.

jharveyb replied:
And that works without this being `*AssetSyncer`?

Roasbeef commented on address/book.go:
```diff
@@ -193,15 +194,70 @@ func NewBook(cfg BookConfig) *Book {
+	log.Debugf("asset %v is unknown, attempting to bootstrap", id.String())
+
+	// Use the AssetSyncer to query our universe federation for the asset.
+	err = b.cfg.Syncer.FetchAssetInfo(ctx, &id)
```
Ah I see here we use sync vs fetch somewhat interchangeably. Perhaps use a `Sync` prefix here since we're reaching out to a remove server to try to obtain the information?

Roasbeef commented on address/book.go:
```diff
@@ -193,15 +194,70 @@ func NewBook(cfg BookConfig) *Book {
+		groupKey = &assetGroup.GroupKey.GroupPubKey
+	}
+
+	err = b.cfg.Syncer.UpdateSyncConfig(ctx, &id, groupKey)
```
I think we can get away with just a single method here that'll try to sync, and if it's able to, it'll add to the config. That way the addr book needs to know less information about syncing/configs, etc.

jharveyb replied:
Was thinking of that initially as well, but it's a bit awkward without adding group lookup to the various universe and federation universes (which I figured we didn't want to do).

From the context of `FetchAssetInfo`, when a sync succeeds, we don't know which proof is for the group anchor, and we would need to use that proof to return results equivalent to what `QueryAssetGroup` returns.

Roasbeef replied:
Not sure I follow: why does it matter which proof gets returned here? IIUC, we just care that we have the genesis information for the asset ID in question. Assuming sync has completed, we should have the genesis info on disk, so we repeat the query that failed to get the info we need.

jharveyb replied:
For a group we also need the group key and the witness from the group anchor to match `QueryAssetGroup`. So those would only be in the group anchor proof.

Or do you mean perform the lookup again in `EnableAssetSync`?

Roasbeef reviewed (changes requested):
Nice work! IMO, this is the syncing model we should've gone with the first time around: only sync stuff on demand when you actually need it. 

Dropped to main comments: 
  1. Looking into moving the query fall back logic deeper into the DB. This can help us cut down on extra DB transactions. 
  2. Only syncing from one server at a time, and simplifying the new duties of the addr book a bit.

GeorgeTsagk commented on address/book.go:
```diff
@@ -193,15 +194,70 @@ func NewBook(cfg BookConfig) *Book {
+
+	// The asset genesis info may have been synced from a universe
+	// server; query for the asset ID again.
+	assetGroup, err = b.cfg.Store.QueryAssetGroup(ctx, id)
```
Could we have `Syncer.FetchAssetInfo` also return some results above? Could save the extra call & store roundtrip

jharveyb replied:
Yes, but contingent on doing the group lookup on the multiverse side / when looking up the root node itself.

GeorgeTsagk commented on tapgarden/custodian_test.go:
```diff
@@ -266,6 +300,80 @@ func TestCustodianNewAddr(t *testing.T) {
+	numAddrs := 1
+	validAddrs := fn.MakeSlice(newAddr)
+
+	// TODO(jhb): remove asset from syncer, assert that address creation
```
what is the expected effect of removing an asset from the syncer?

jharveyb replied:
Here, the `address.Book` should still succeed to make an address, as `QueryAssetGroup` should find the asset genesis (since the address should have been imported after the `AddAsset` call above.

GeorgeTsagk commented on itest/addrs_test.go:
```diff
@@ -184,6 +187,151 @@ func testMultiAddress(t *harnessTest) {
+
+	// Bob should not be able to make an address for a random asset ID
+	// that he nor Alice are aware of.
+
```
nit: extra newline

GeorgeTsagk commented on itest/addrs_test.go:
```diff
@@ -184,6 +187,151 @@ func testMultiAddress(t *harnessTest) {
+
+	// Bob should be able to make an address for both new assets minted by
+	// Alice, even though he has not synced their issuance proofs.
+
```
nit: extra nl

GeorgeTsagk reviewed (commented):
nice PR! had the chance to dive into a lot of code while reading, JIT syncing looks promising

Roasbeef commented:
Removed from draft so the itests can run as well.

jharveyb commented on universe/auto_syncer.go:
```diff
@@ -531,6 +518,137 @@ func (f *FederationEnvoy) SetAllowPublicAccess() error {
+		return err
+	}
+
+	syncDiffs := fn.Collect(returnedSyncDiffs)
```
Yeah I think once the caller syncs the full state from any federation syncer, the other calls would have another path to exit early from the local and remote root matching. But also not sure.

jharveyb commented on rpcserver.go:
```diff
@@ -2993,6 +2995,109 @@ func (r *rpcServer) QueryAssetRoots(ctx context.Context,
+	}
+
+	// If an asset group was found, fetch the roots for that group.
+	if foundGroupKey != nil {
```
Yup considered that initially but I think current version simplifies the err handling a bit, it works out.

jharveyb commented:
Reworked the config for disabling the asset syncer. itest extended to test disabling the asset syncer + asserts that the config update succeeds + tests the syncer with a grouped asset. Also refactored the itest harness a bit to make adding future args for test setup a bit easier.

Roasbeef commented on rpcserver.go:
```diff
@@ -2974,21 +3046,21 @@ func (r *rpcServer) QueryAssetRoots(ctx context.Context,
 	}
 
-	transferRootRPC, err := marshalUniverseRoot(transferRoot)
+	transferRootRPC, err = marshalUniverseRoot(transferRoot)
```
Won't this panic if nothing was found above?

jharveyb replied:
Hmm yeah I think the `RootNode` call above would err,  return a real root, or `Root{}`. But `marshalUniverseRoot has an empty nil-ness check that exits early on `Root{}`:

https://github.com/lightninglabs/taproot-assets/blob/e569cf553d324297c303d190678320fcc50b9c10/rpcserver.go#L2751

 And we already hit this as the unknown assets would have no transfer root.

Roasbeef commented on rpcserver.go:
```diff
@@ -2957,13 +2957,85 @@ func (r *rpcServer) QueryAssetRoots(ctx context.Context,
 	if err != nil {
+		return nil, err
+	}
+
```
If we have an actual root here, then shouldn't we just return early rather than relying on the default behavior below?

jharveyb replied:
We can; in practice, `mayBeGrouped` would be false, so we skip the group lookup.

Looked it and early exit does reduce indention though, so switched to that.

Roasbeef commented on universe/auto_syncer.go:
```diff
@@ -531,6 +518,134 @@ func (f *FederationEnvoy) SetAllowPublicAccess() error {
+
+		err := fmt.Errorf("unable to fetch set of universe servers: %v",
+			err)
+		f.reportErr(err)
```
We should remove this (depends on order of which PRs land first), as is it'll make the entire daemon shutdown if we can't read the set of servers for w/e reason. So we should just log a `Warnf` here.

jharveyb replied:
K will mirror what you have up.

Roasbeef commented on tapcfg/config.go:
```diff
@@ -253,6 +253,12 @@ type UniverseConfig struct {
+// AddressConfig is the config that houses any address Book related config
+// values.
+type AddrBookConfig struct {
+	DisableSyncer bool `long:"disable-syncer" description:"If true, tapd will not try to sync issuance proofs for unknown assets when creating an address."`
```
👍

Roasbeef commented on itest/tapd_harness.go:
```diff
@@ -80,12 +80,28 @@ type tapdConfig struct {
+	addrAssetSyncerDisable  bool
+}
+
+type harnessOption func(*harnessOpts)
```
👍

Roasbeef reviewed (approved):
LGTM 🎡

Just one lingering questions about early return and if we want to return an error on unknown root, or just serialize an empty proto.

jharveyb commented:
Looks like new flake on Postgres?

dstadulis commented:
itests: if you modify the global state of [something] it'll cause dysfunction.  FIxes are known and will arrive today

jharveyb commented:
Ready for final 2nd review.

GeorgeTsagk commented on itest/addrs_test.go:
```diff
@@ -184,6 +190,243 @@ func testMultiAddress(t *harnessTest) {
+
+	// We'll now add the main node, as a member of Bob's Universe
+	// federation. We expect that their state is synchronized shortly after
+	// the call returns.
```
nit: We haven't had any flakes here, but should we wait with `Eventualy` to ensure nodes are sync'd instead?

GeorgeTsagk commented on itest/addrs_test.go:
```diff
@@ -184,6 +190,243 @@ func testMultiAddress(t *harnessTest) {
+	// Restart Bob again with the syncer enabled. Bob should be able to make
+	// an address for both new assets minted by Alice, even though he has
+	// not synced their issuance proofs.
+	restartBobNoUniSync(false)
```
:+1: 
Last note w.r.t being able to enable/disable JIT sync: Is the "global sync" (fetch all) and JIT sync going to live side-by-side in the future? Feels to me that if JIT is disabled then global sync must be present, and if JIT is enabled global has to be off, as these two sort of overlap

GeorgeTsagk replied:
does that lead us to 2 future modes for sync?

jharveyb replied:
Good Q: So if global sync is on, the syncer setting sort of doesn't matter; the only way to not know about an asset would be if it was pushed to a federation member after your last sync (up to 10 mins ago).

The default with this PR and the others in main is "JIT sync on, global sync off". With manual global sync still possible.

With JIT on anything you sync from a global sync, you haven't interacted with yet, so I think that's a reasonable default.

GeorgeTsagk reviewed (approved):
LGTM :fire: big step for smart sync :eyes:
